### PR TITLE
Bug 5904 minimum order size rejected order

### PR DESCRIPTION
**Pull request by Marinovsky**
Update SymbolsPropertiesDB and Crypto Brokerage Models (QuantConnect#5904)
#### Description

- Update symbols-properties-database.csv with a minimum order size parameter for crypto assets
- Add tests to check if the the SymbolsPropertiesDatabase is loading properly the Minimum Order Size data
- Update Binance, Bitfinex, Kraken and GDAX brokerages models in order to take into account the order size when placing or updating an order
- Create new method IsOrderSizeValid() in default brokerage model in order to check if the order size is valid
- Create new class TestHelpers() to provide useful methods for the brokerages tests
- Add BTCUSD Bitfinex hourly historical data from October 2013 to check Bitfinex brokerage model
- Add tests to check if Binance, Bitfinex, Kraken and GDAX models are taking into account the order size when placing an order
- Add regression test to check if Bitfinex brokerage model is taking into account the order size when placing or updating an order
#### Related Issue

Closes https://github.com/QuantConnect/Lean/issues/5904

#### Motivation and Context

With this changes the helper methods such as SetHoldings and CalculateOrderSize should not result in a quantity that will be rejected by the brokerage minimum order size.
#### Requires Documentation Change

This changes don't require changes or updates to documentation
#### How Has This Been Tested?

- To check the Symbol Properties Data base was loading properly the minimum order size parameter we assert the value is loading with a the value of a random crypto asset extracted manually from the symbol-properties-database.csv
- To check the crypto brokerages models were taking into account the order size when submiting an order we assert an order with a size less than the allowed and an order with a valid size
- To check the behavior at a system level we write a regression algoritm that first tries to place an order with an invalid size, then tries to update an order with an order with an invalid size and finally places and update an order with a valid size
#### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Algorithm.CSharp/BuyBTCWithLessThanOneDollarAlgorithm.cs [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+using System;
+using System.Collections.Generic;
+using QuantConnect.Orders;
+using QuantConnect.Interfaces;
+using QuantConnect.Data;
+
+namespace QuantConnect.Algorithm.CSharp
+{
+    public class BuyBTCWithLessThanOneDollarAlgorithm : QCAlgorithm, IRegressionAlgorithmDefinition
```
Comment (Martin-Molinero) [inline Algorithm.CSharp/BuyBTCWithLessThanOneDollarAlgorithm.cs:24]: - `BuyBTCWithLessThanOneDollarAlgorithm` let's rename him to something like `MinimumOrderSizeRegressionAlgorithm`
- Let's add a xml summary description for the algorithm `...This algorithm asserts that the minimum order size is respected..etc`

File: Algorithm.CSharp/BuyBTCWithLessThanOneDollarAlgorithm.cs [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    {
+        public override void Initialize()
+        {
+            SetStartDate(2013, 10, 1);
+            SetEndDate(2013, 10, 1);
+            SetCash(1);
```
Comment (Martin-Molinero) [inline Algorithm.CSharp/BuyBTCWithLessThanOneDollarAlgorithm.cs:30]: >  SetCash(1);

Why? Let's remove this if not required, the simpler 10x better 👍

File: Algorithm.CSharp/BuyBTCWithLessThanOneDollarAlgorithm.cs [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+            SetBrokerageModel(Brokerages.BrokerageName.Bitfinex, AccountType.Cash);
+            AddCrypto("BTCUSD", Resolution.Minute, Market.Bitfinex);
```
Comment (Martin-Molinero) [inline Algorithm.CSharp/BuyBTCWithLessThanOneDollarAlgorithm.cs:32]: > Market.Bitfinex

Specifying the market shouldn't be required, let's remove it 👌. Default resolution is already Minute too

File: Algorithm.CSharp/BuyBTCWithLessThanOneDollarAlgorithm.cs [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+                    Quantity = 0.002m
+                });
+
+                if (!validUpdate.IsSuccess)
+                {
+                    throw new Exception("Update was expected to succeed");
+                }
+
+                // Order to fill the portfolio
+                var closeOrder = MarketOrder("BTCUSD", 0.0002);
```
Comment (Martin-Molinero) [inline Algorithm.CSharp/BuyBTCWithLessThanOneDollarAlgorithm.cs:71]: > var closeOrder =

Is not required?

File: Common/Brokerages/BinanceBrokerageModel.cs [hunk 5]
```diff
@@ -88,6 +91,50 @@ public override IFeeModel GetFeeModel(Security security)
...
         }
 
+        /// <summary>
+        /// Binance does not support update of orders
+        /// </summary>
+        /// <param name="security"></param>
+        /// <param name="order"></param>
+        /// <param name="request"></param>
+        /// <param name="message"></param>
+        /// <returns></returns>
```
Comment (Martin-Molinero) [inline Common/Brokerages/BinanceBrokerageModel.cs:101]: Missing docs, let's review all the changes for missing docs 👍 

> using static QuantConnect.StringExtensions; & using System.Linq;

Not required, let's review all changes for none required imports

File: Common/Brokerages/BinanceBrokerageModel.cs [hunk 6]
```diff
@@ -88,6 +91,50 @@ public override IFeeModel GetFeeModel(Security security)
...
+        /// <remarks>
+        /// For example, a brokerage may have no connectivity at certain times, or an order rate/size limit
+        /// </remarks>
+        /// <param name="security"></param>
+        /// <param name="order">The order to be processed</param>
+        /// <param name="message">If this function returns false, a brokerage message detailing why the order may not be submitted</param>
+        /// <returns>True if the brokerage could process the order, false otherwise</returns>
+        public override bool CanSubmitOrder(Security security, Order order, out BrokerageMessageEvent message)
+        {
+            if (!IsValidOrderSize(security, order.AbsoluteQuantity, out message))
```
Comment (Martin-Molinero) [inline Common/Brokerages/BinanceBrokerageModel.cs:121]: Cool! `order.AbsoluteQuantity` can just send quantity since `IsValidOrderSize` is already calling Math.Abs on it, which seems like the right think 👌

File: Common/Brokerages/GDAXBrokerageModel.cs [hunk 7]
```diff
@@ -152,14 +110,8 @@ public override bool CanSubmitOrder(Security security, Order order, out Brokerag
                 return false;
             }
 
-            decimal minimumOrderSize;
-            if (MinimumOrderSizes.TryGetValue(security.Symbol.Value, out minimumOrderSize) &&
-                Math.Abs(order.Quantity) < minimumOrderSize)
+            if(!IsValidOrderSize(security, order.AbsoluteQuantity, out message))
```
Comment (Martin-Molinero) [inline Common/Brokerages/GDAXBrokerageModel.cs:113]: Sweet!

File: Common/Brokerages/KrakenBrokerageModel.cs [hunk 8]
```diff
@@ -94,20 +115,6 @@ public override bool CanSubmitOrder(Security security, Order order, out Brokerag
...
-        
-        /// <summary>
-        /// Kraken does no support update of orders
-        /// </summary>
-        /// <param name="security">Security</param>
-        /// <param name="order">Order that should be updated</param>
-        /// <param name="request">Update request</param>
-        /// <param name="message">Outgoing message</param>
-        /// <returns>Always false as Kraken does no support update of orders</returns>
-        public override bool CanUpdateOrder(Security security, Order order, UpdateOrderRequest request, out BrokerageMessageEvent message)
```
Comment (Martin-Molinero) [inline Common/Brokerages/KrakenBrokerageModel.cs:106]: Let's revert this change since it's not required, this version even has the missing docs 👍

File: Engine/TransactionHandlers/BrokerageTransactionHandler.cs [hunk 9]
```diff
@@ -319,6 +319,10 @@ public OrderTicket UpdateOrder(UpdateOrderRequest request)
                 var order = GetOrderByIdInternal(request.OrderId);
                 var orderQuantity = request.Quantity ?? ticket.Quantity;
 
+                // ensure the order is tagged with a currency
+                var security = _algorithm.Securities[order.Symbol];
+                BrokerageMessageEvent message;
```
Comment (Martin-Molinero) [inline Engine/TransactionHandlers/BrokerageTransactionHandler.cs:324]: Unrequired changes?

File: Engine/TransactionHandlers/BrokerageTransactionHandler.cs [hunk 10]
```diff
@@ -362,6 +366,8 @@ public OrderTicket UpdateOrder(UpdateOrderRequest request)
                 {
                     request.SetResponse(OrderResponse.Success(request), OrderRequestStatus.Processing);
                     _orderRequestQueue.Add(request);
+
+                    WaitForOrderSubmission(ticket);
```
Comment (Martin-Molinero) [inline Engine/TransactionHandlers/BrokerageTransactionHandler.cs:370]: Don't think this PR should be changing this?

File: Tests/Common/Brokerages/BinanceBrokerageModelTests.cs [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+using QuantConnect.Tests.Brokerages;
+using QuantConnect.Orders;
+
+namespace QuantConnect.Tests.Common.Brokerages
+{
+
+    [TestFixture, Parallelizable(ParallelScope.All)]
+    public class BinanceBrokerageModelTests
+    {
+        private readonly BinanceBrokerageModel _unit = new BinanceBrokerageModel();
```
Comment (Martin-Molinero) [inline Tests/Common/Brokerages/BinanceBrokerageModelTests.cs:28]: `_unit` let's rename this guy 👍

File: Tests/Common/Brokerages/BinanceBrokerageModelTests.cs [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+        [TestCase(0.01, true)]
+        [TestCase(0.000009, false)]
+        public void CanSubmitOrder_WhenQuantityIsLargeEnough(decimal orderQuantity, bool isValidOrderQuantity)
+        {
+            BrokerageMessageEvent message;
+            var order = new Mock<Order>();
+
+            order.Object.Quantity = orderQuantity;
+
+            Assert.AreEqual(isValidOrderQuantity, _unit.CanSubmitOrder(TestsHelpers.GetSecurity(symbol: "BTCEUR", market: Market.Binance, quoteCurrency: "EUR"), order.Object, out message));
```
Comment (Martin-Molinero) [inline Tests/Common/Brokerages/BinanceBrokerageModelTests.cs:39]: cool! 👍

File: Data/symbol-properties/symbol-properties-database.csv [hunk 13]
```diff
@@ -81,9 +81,7 @@ oanda,XPDUSD,cfd,Palladium,USD,1,0.001,1
 oanda,XPTUSD,cfd,Platinum,USD,1,0.001,1
 
 usa,[*],option,,USD,100,0.01,1
-
 usa,[*],equity,,USD,1,0.01,1
-
```
Comment (Martin-Molinero) [inline Data/symbol-properties/symbol-properties-database.csv:86]: Unrequired changes, let's revert 👍

File: Common/Securities/SymbolProperties.cs [hunk 14]
```diff
@@ -72,10 +72,18 @@ public string MarketTicker
             get;
         }
 
+        /// <summary>
+        /// The minimum order size allowed
+        /// </summary>
```
Comment (Martin-Molinero) [inline Common/Securities/SymbolProperties.cs:77]: For pairs like forex or crypto we should specify in the docs in what unit is this expected to be expressed in, believe this is in the Base currency, right?

File: Data/symbol-properties/symbol-properties-database.csv [hunk 15]
```diff
@@ -365,2332 +363,2330 @@ oanda,USDZAR,forex,USD/ZAR,ZAR,1,0.00001,1
...
+binance,C98BNB,crypto,C98BNB,BNB,1,0.00000010,1.00000000,C98BNB,0.10000000
+binance,C98BTC,crypto,C98BTC,BTC,1,0.00000001,1.00000000,C98BTC,0.10000000
+binance,C98BUSD,crypto,C98BUSD,BUSD,1,0.00001000,1.00000000,C98BUSD,0.10000000
+binance,C98USDT,crypto,C98USDT,USDT,1,0.00001000,1.00000000,C98USDT,0.10000000
+binance,CAKEBNB,crypto,CAKEBNB,BNB,1,0.00001000,0.10000000,CAKEBNB,0.01000000
+binance,CAKEBRL,crypto,CAKEBRL,BRL,1,0.01000000,0.01000000,CAKEBRL,0.01000000
+binance,CAKEBTC,crypto,CAKEBTC,BTC,1,0.00000010,0.01000000,CAKEBTC,0.01000000
+binance,CAKEBUSD,crypto,CAKEBUSD,BUSD,1,0.00100000,0.00100000,CAKEBUSD,0.01000000
+binance,CAKEGBP,crypto,CAKEGBP,GBP,1,0.00100000,0.00100000,CAKEGBP,0.01000000
+binance,CAKEUSDT,crypto,CAKEUSDT,USDT,1,0.00100000,0.00100000,CAKEUSDT,0.01000000
```
Comment (Martin-Molinero) [inline Data/symbol-properties/symbol-properties-database.csv:1111]: We have some weird ending zeros in some of these, think we can lose them? 😁

**Review by Martin-Molinero — CHANGES_REQUESTED** [review]
Awesome! leaving a few minor requests

**Review by Martin-Molinero — APPROVED** [review]
Sweet! last minor comments

**Review by Martin-Molinero — APPROVED** [review]

**Review by jaredbroad — APPROVED** [review]
